lib/index.js

```diff
@@ -183,3 +184,203 @@ exports.directory = FileSystem.directory;
+
+const mapping = {};
+
+mapping.createContext = ({output, options = {}, target}, newContext) =>
+  Object.assign(
+    {
+      // Assign options and set defaults if needed
+      options: {
+        recursive: options.recursive !== false,
+        lazyLoad: options.lazyLoad !== false
+      },
+      output,
+      target
+    },
+    newContext
+  );
+
+mapping.addFile = function(context, stats, isRoot) {
+  const {output, target} = context;
+  const {lazyLoad} = context.options;
+
+  if (!stats.isFile()) {
+    throw new Error(`${target} is not a valid file!`);
+  }
+
+  const outputPropKey = isRoot ? target : path.win32.basename(target);
+
+  output[outputPropKey] = () => {
+    const content = !lazyLoad ? fs.readFileSync(target) : '';
+    const file = exports.file(Object.assign({}, stats, {content}))();
+
+    if (lazyLoad) {
+      Object.defineProperty(file, '_content', {
+        get() {
+          const res = exports.bypass(() => fs.readFileSync(target));
+          Object.defineProperty(file, '_content', {
+            value: res,
+            writable: true
+          });
+          return res;
+        },
+        set(data) {
+          Object.defineProperty(file, '_content', {
+            value: data,
+            writable: true
+          });
+        },
+        configurable: true
+      });
+    }
+
+    return file;
+  };
+
+  return output[outputPropKey];
+};
+
+mapping.addDir = function(context, stats, isRoot) {
+  const {target, output} = context;
+  const {recursive} = context.options;
+
+  if (!stats.isDirectory()) {
+    throw new Error(`${target} is not a valid directory!`);
+  }
+
+  stats = Object.assign({}, stats);
+  const outputPropKey = isRoot ? target : path.win32.basename(target);
+
+  // On windows platforms, directories do not have the executable flag, which causes FileSystem.prototype.getItem
+  // to think that the directory cannot be traversed. This is a workaround, however, a better solution may be to
+  // re-think the logic in FileSystem.prototype.getItem
+  // This workaround adds executable privileges if read privileges are found
+  stats.mode = fixWin32Permissions(stats.mode);
+
+  // Create directory factory
+  const directoryItems = {};
+  output[outputPropKey] = exports.directory(
+    Object.assign(stats, {items: directoryItems})
+  );
+
+  fs.readdirSync(target).forEach(p => {
+    const absPath = path.join(target, p);
+    const stats = fs.statSync(absPath);
+    const newContext = mapping.createContext(context, {
+      target: absPath,
+      output: directoryItems
+    });
+
+    if (recursive && stats.isDirectory()) {
+      mapping.addDir(newContext, stats);
+    } else if (stats.isFile()) {
+      mapping.addFile(newContext, stats);
+    }
+  });
+
+  return output[outputPropKey];
+};
+
+mapping.fixupPath = p => {
+  if (typeof p !== 'string') {
+    throw new TypeError(`Invalid path. All paths must be strings`);
+  }
+  return path.resolve(p);
```

Comment:
`path.resolve(p)` already throws a `TypeError` if the argument is not a string.  I'm curious why this is needed.

Reply:
Simply to provide a specific error message that's makes the issue easily understood. Throwing custom TypeError with a more understandable message is a standard practice. I can remove if you'd prefer.

Reply:
Yes, I'd prefer to have the error that people expect from `path.resolve()`.